zookeeper-server/src/main/java/org/apache/zookeeper/server/quorum/MultipleAddresses.java

```diff
@@ -0,0 +1,226 @@
+    }
+
+    /**
+     * Returns all ports
+     *
+     * @return list of all ports
+     */
+    public List<Integer> getAllPorts() {
+        return addresses.stream().map(InetSocketAddress::getPort).distinct().collect(Collectors.toList());
+    }
+
+    /**
+     * Returns distinct list of all host strings
+     *
+     * @return list of all hosts
+     */
+    public List<String> getAllHostStrings() {
+        return addresses.stream().map(InetSocketAddress::getHostString).distinct().collect(Collectors.toList());
+    }
+
+    public void addAddress(InetSocketAddress address) {
+        addresses.add(address);
+    }
+
+    /**
+     * Returns reachable address. If none is reachable than throws exception.
+     *
+     * @return address which is reachable.
+     * @throws NoRouteToHostException if none address is reachable
+     */
+    public InetSocketAddress getReachableAddress() throws NoRouteToHostException {
+        AtomicReference<InetSocketAddress> address = new AtomicReference<>(null);
+        getInetSocketAddressStream().forEach(addr -> checkIfAddressIsReachableAndSet(addr, address));
+
+        if (address.get() != null) {
+            return address.get();
+        } else {
+            throw new NoRouteToHostException("No valid address among " + addresses);
+        }
+    }
+
+    /**
+     * Returns reachable address or first one, if none is reachable.
+     *
+     * @return address which is reachable or fist one.
+     */
+    public InetSocketAddress getReachableOrOne() {
+        InetSocketAddress address;
+        try {
+            address = getReachableAddress();
+        } catch (NoRouteToHostException e) {
+            address = getOne();
+        }
+        return address;
+    }
+
+    /**
+     * Performs a DNS lookup for addresses.
+     *
+     * If the DNS lookup fails, than address remain unmodified.
+     */
+    public void recreateSocketAddresses() {
+        Set<InetSocketAddress> temp = Collections.newSetFromMap(new ConcurrentHashMap<>());
+        temp.addAll(getInetSocketAddressStream().map(this::recreateSocketAddress).collect(Collectors.toSet()));
+        addresses = temp;
+    }
+
+    /**
+     * Returns first address from set.
+     *
+     * @return address from a set.
+     */
+    public InetSocketAddress getOne() {
+        return addresses.iterator().next();
+    }
+
+    private void checkIfAddressIsReachableAndSet(InetSocketAddress address,
+                                                 AtomicReference<InetSocketAddress> reachableAddress) {
+        for (int i = 0; i < 5 && reachableAddress.get() == null; i++) {
+            try {
+                if (address.getAddress().isReachable((i + 1) * timeout)) {
+                    reachableAddress.compareAndSet(null, address);
+                    break;
+                }
+                Thread.sleep(timeout);
+            } catch (NullPointerException | IOException | InterruptedException ignored) {
```

Comment:
catching NullPointerException is a code and usually a bad practice.
please remove this case

catching InterruptedException requires to reset Thread#interrupt flag, please fix or explain

Reply:
Thanks, I agree... this one needs to be refactored. I will remove the atomic reference and do a simple parallel stream filtering in the `getReachableAddress()`. Please review after I push the new commit.

Reply:
@symat why are you talking about parallelStream ? we are not talking about hundreds of addresses, using more that one thread won't help and maybe it could be more expensive

Reply:
well, we are using parallel streams only in two places, where IO operations happen (one for hostname resolution, one for checking if the address is reachable). I agree that we don't have hundreds of addresses, but even with a few addresses, parallelizing can mean that we can find a reachable address for a remote Quorum Server a few seconds quicker.

I agree that it is not a huge deal... but it is a trivial optimization in the code and I don't think it would be unsafe in any ways.

I cleaned up this part in my last commit. If you still think it is unnecessary and would be better to use simple streams, then we can go for it. (You are right that in the rare case of a network error that few seconds will not matter much)